pkg/volume/csi/csi_mounter.go

```diff
@@ -206,31 +206,30 @@ func (c *csiMountMgr) SetUpAt(dir string, fsGroup *int64) error {
 	}
 
 	// apply volume ownership
-	if !c.readOnly && fsGroup != nil {
-		err := volume.SetVolumeOwnership(c, fsGroup)
-		if err != nil {
-			// attempt to rollback mount.
-			glog.Error(log("mounter.SetupAt failed to set fsgroup volume ownership for [%s]: %v", c.volumeID, err))
-			glog.V(4).Info(log("mounter.SetupAt attempting to unpublish volume %s due to previous error", c.volumeID))
-			if unpubErr := csi.NodeUnpublishVolume(ctx, c.volumeID, dir); unpubErr != nil {
-				glog.Error(log(
-					"mounter.SetupAt failed to unpublish volume [%s]: %v (caused by previous NodePublish error: %v)",
-					c.volumeID, unpubErr, err,
-				))
-				return fmt.Errorf("%v (caused by %v)", unpubErr, err)
-			}
+	// The following logic is derived from https://github.com/kubernetes/kubernetes/issues/66323
+	// if fstype is "", then skip fsgroup (could be indication of non-block filesystem)
+	// if fstype is provided and pv.AccessMode == ReadWriteOnly, then apply fsgroup
 
-			if unmountErr := removeMountDir(c.plugin, dir); unmountErr != nil {
-				glog.Error(log(
-					"mounter.SetupAt failed to clean mount dir [%s]: %v (caused by previous NodePublish error: %v)",
-					dir, unmountErr, err,
-				))
-				return fmt.Errorf("%v (caused by %v)", unmountErr, err)
-			}
+	err = applyFSGroup(c, fsType, fsGroup)
+	if err != nil {
+		// attempt to rollback mount.
+		glog.Error(log("mounter.SetupAt failed to set fsgroup volume ownership for [%s]: %v", c.volumeID, err))
+		if unpubErr := csi.NodeUnpublishVolume(ctx, c.volumeID, dir); unpubErr != nil {
+			glog.Error(log(
```

Comment:
Still not a big fan of logging the error that we're also going to return.  Can we generate a new error message that includes the previous error and just return that?

Reply:
@msau42  I think in certain context logging the error is OK specially to add `context` to the error being returned. This nested case is tricky since there are two errors being considered. Here I am returning the second error that caused the return while logging both to give context... 

What I am not a fan of is using the error value as a log message and hope the caller will do a good job explaining what happened. Let me see what I can do here.

Reply:
Can you just have the error message contain all the context?

Reply:
@msau42 I made suggested changes (test case and log clean up).  Please take a look! 
Thank you for your help.